### PR TITLE
feat(architecture): replace DrivingPolicy with TrajectoryPlanner and increase BEV resolution to 450x300

### LINKED ISSUE #35: Proposal: Increase BEV Feature Map Resolution
## Context

The current BEV feature map resolution is 7×7, which comes directly from the backbone's final feature map size (256×256 input → 7×7 after 5 stages of downsampling). This was never intentionally designed as the BEV grid resolution — it's just the default backbone output that gets passed through.

At 7×7, if we assume a perception range of ~50m around the vehicle, each cell covers roughly 7m × 7m. A single passenger car (4.5m long) fits entirely within one cell, making it impossible to distinguish spatial relationships between nearby objects.

## How other models handle this

Most published BEV perception models define an explicit BEV grid independent of the backbone output resolution, then use attention or projection to populate it:

| Model | BEV Resolution | Perception Range | Cell Size | Reference |
|-------|---------------|-----------------|-----------|-----------|
| BEVFormer | 200×200 | 51.2m × 51.2m | 0.256m | [Li et al., ECCV 2022](https://arxiv.org/abs/2203.17270) |
| UniAD | 200×200 | 51.2m × 51.2m | 0.256m | [Hu et al., CVPR 2023](https://arxiv.org/abs/2212.10156) |
| BEVDet | 128×128 | 51.2m × 51.2m | 0.4m | [Huang et al., 2022](https://arxiv.org/abs/2112.11790) |
| BEVDepth | 128×128 | 51.2m × 51.2m | 0.4m | [Li et al., AAAI 2023](https://arxiv.org/abs/2206.10092) |
| PETR v2 | 100×100 | 51.2m × 51.2m | 0.512m | [Liu et al., ICCV 2023](https://arxiv.org/abs/2206.01256) |
| VAD | 200×200 | 60m × 30m | 0.3m / 0.15m | [Jiang et al., ICCV 2023](https://arxiv.org/abs/2303.12077) |
| SparseBEV | 200×200 | 51.2m × 51.2m | 0.256m | [Liu et al., ICCV 2023](https://arxiv.org/abs/2308.09244) |
| StreamPETR | 100×100 | 51.2m × 51.2m | 0.512m | [Wang et al., ICCV 2023](https://arxiv.org/abs/2303.11926) |

The common range is **100×100 to 200×200**, giving cell sizes of **0.25m to 0.5m**. This is fine enough to distinguish individual vehicles, lane markings, and pedestrians.

## Proposed Approach

Define an explicit BEV grid in the BEV fusion module, decoupled from backbone output:

```python
# Instead of using backbone's 7×7 directly:
bev_h, bev_w = 100, 100  # or 200×200
perception_range = 51.2  # meters in each direction

# Learnable BEV queries
bev_queries = nn.Parameter(torch.randn(bev_h * bev_w, embed_dim))

# Use cross-attention to populate from multi-view features
# (camera_params tells the attention where each BEV query maps to in each camera)
```

## Trade-offs to consider

| Resolution | Cell Size (50m range) | VRAM (embed_dim=256, fp16) | Compute |
|-----------|----------------------|---------------------------|---------|
| 50×50 | 1.0m | ~3 MB | Low |
| 100×100 | 0.5m | ~13 MB | Medium |
| 200×200 | 0.25m | ~50 MB | High |

Given that AutoE2E targets real-time inference on automotive hardware, **100×100 might be a reasonable starting point** — 0.5m resolution is enough for trajectory planning without the VRAM cost of 200×200.

## Questions

- What perception range should we target? 50m? 100m forward-biased (e.g., 80m front, 20m rear)?
- Should we start with 100×100 and see if it's sufficient for trajectory prediction, or go straight to 200×200?
- Does this interact with #23 (embed_dim unification)? A smaller embed_dim (256 vs 1440) would make higher BEV resolution more feasible.

Would appreciate thoughts on what makes sense for our use case and hardware targets.

### PR DESCRIPTION
**Pull request by riita10069**
## Summary

Resolves #35 — Replaces the flatten-then-MLP `DrivingPolicy` with a `TrajectoryPlanner` that uses ego query + GRU + deformable cross-attention to BEV features, and increases BEV resolution from 7×7 to 450×300 with a forward-biased asymmetric perception range.

## Motivation

The previous `DrivingPolicy` flattened the BEV feature map into a single vector before an MLP. This approach cannot scale beyond small BEV grids — at 450×300×256, flattening would produce 34.5M input dimensions. The literature (UniAD, VAD, BEVFormer) consistently uses ego query + cross-attention for this reason.

The 450×300 grid with 0.4m cell size and forward-biased range (120m front, 60m rear, 60m left/right) was chosen per discussion in #35 to support safety-critical high-speed driving scenarios.

### New: `TrajectoryPlanner` (replaces `DrivingPolicy`)
- Learnable ego query + GRU autoregressive decoder
- At each of 64 timesteps, the GRU hidden state cross-attends to BEV features via deformable attention (`F.grid_sample` — same pattern as `BEVViewFusion`)
- Outputs: `trajectory (B, 128)` + `ego_hidden (B, 256)`
- BEV resolution-agnostic: works with any spatial size

### BEV Resolution
- `BEVViewFusion` default: `bev_h=450, bev_w=300`
- `pc_range=(-60, -60, -5, 120, 60, 3)` — X: -60m to +120m (front-biased), Y: ±60m
- Cell size: 0.4m
- Fully configurable via `view_fusion_kwargs`

### `FutureState`
- Now consumes `ego_hidden (256-dim)` instead of the abolished 14-dim compressed visual vector
- Ego intent injected as per-channel bias over BEV spatial dims

### `AutoE2E.forward()` signature change
```python
# Before:
trajectory, compressed_visual, future = model(x, visual_history, egomotion_history, ...)

# After:
trajectory, ego_hidden, future = model(x, egomotion_history, ...)
```
- `visual_history` input removed
- `compressed_visual_feature_vector` output removed
- `ego_hidden` replaces both

## Files Changed (17 commits)

| Component | Change |
|-----------|--------|
| `trajectory_planner.py` | New module |
| `auto_e2e.py` | DrivingPolicy → TrajectoryPlanner |
| `feature_fusion.py` | Removed 8×8 hardcoding, `image_feature_size` + `view_fusion_kwargs` params |
| `view_fusion/bev_fusion.py` | Configurable default 450×300 |
| `view_fusion/__init__.py` | Fixed stale `embed_dim=1440` default |
| `future_state.py` | ego_hidden (256-dim) input |
| `driving_policy.py` | Deleted |
| `tests/conftest.py` | Mock wired to new architecture (8×8 BEV for speed) |
| `tests/test_auto_e2e.py` | Full rewrite — 101 tests |
| `speed_benchmark.py` | Updated to new forward signature |
| `inference/run_forward_pass.py` | Updated |
| `nvidia_physical_ai/dataset.py` | Removed `visual_history` |
| `nvidia_physical_ai/forward_pass_test.py` | Updated |

## Test plan

- [x] 101 unit tests pass (CPU, mock backbone with 8×8 BEV)
- [x] All fusion modes (concat, cross_attn, bev) pass
- [x] BEV 450×300 default resolution verified via shape assertion
- [x] Asymmetric BEV grid test
- [x] TrajectoryPlanner component tests (arbitrary resolution, gradient flow, egomotion influence)
- [x] FutureState accepts ego_hidden
- [x] NumViews flexibility (1, 4, 8, 12 views)

## Known considerations

- The 64-step autoregressive GRU loop is sequential — a known latency concern for future optimization (non-autoregressive decoding)
- `FutureState.predict_future_2` at full 450×300 resolution outputs ~550MB/sample in fp32; mixed precision or spatial downsampling may be needed for training
- `offset_scale=0.1` limits sampling offsets to 10% of BEV grid from reference point; long-range attention is still possible via the reference point itself

### CODE REVIEW

**Comment by riita10069** [comment]
## Note: FutureState memory cost at full BEV resolution

`FutureState.predict_future_2` outputs `(B, 4*embed_dim, bev_h, bev_w)` = `(B, 1024, 450, 300)` at the default grid size. Per-sample memory:

- Output tensor alone: ~550 MB (fp32)
- With autograd saved tensors + Conv2d intermediate activations: ~1.5 GB/sample
- batch_size=4 → ~6 GB for this layer alone (before optimizer state)

This is trainable on 80GB HBM (A100/B200) but not on 16GB GPUs (T4/g4dn).

**Possible mitigations (for later):**
- Spatial downsample before FutureState (e.g., `AdaptiveAvgPool2d(32, 32)` on BEV features before passing to FutureState)
- Reduce output channels (sequential per-timestep prediction instead of 4× channel blow-up)
- Gradient checkpointing on this module
- Start training at smaller BEV resolution (100×100) and scale up

Not blocking for this PR since no training loop exists yet (#14), but should be addressed before actual training begins.

**Review by m-zain-khawaja — APPROVED** [review]
Approved - thanks @riita10069
